graph/src/components/ethereum/adapter.rs

```diff
@@ -943,63 +942,80 @@ pub fn blocks_with_triggers(
+        .concat2()
+        .join(
+            adapter
+                .clone()
+                .block_hash_by_block_number(&logger, chain_store.clone(), to, true)
+                .then(move |to_hash| match to_hash {
+                    Ok(n) => n.ok_or_else(|| {
+                        warn!(logger2,
+                                "Ethereum endpoint is behind";
+                                "url" => eth_clone.url_hostname()
+                        );
+                        format_err!("Block {} not found in the chain", to)
                     }),
-            )
-            .map(move |(triggers, to_hash)| {
-                let mut block_hashes: HashSet<H256> =
-                    triggers.iter().map(EthereumTrigger::block_hash).collect();
-                let mut triggers_by_block: HashMap<u64, Vec<EthereumTrigger>> =
-                    triggers.into_iter().fold(HashMap::new(), |mut map, t| {
-                        map.entry(t.block_number()).or_default().push(t);
-                        map
-                    });
-
-                debug!(logger, "Found {} relevant block(s)", block_hashes.len());
-
-                // Make sure `to` is included, even if empty.
-                block_hashes.insert(to_hash);
-                triggers_by_block.entry(to).or_insert(Vec::new());
+                    Err(e) => Err(e),
+                }),
+        )
+        .compat()
+        .await?;
+
+    let mut block_hashes: HashSet<H256> =
+        triggers.iter().map(EthereumTrigger::block_hash).collect();
+    let mut triggers_by_block: HashMap<u64, Vec<EthereumTrigger>> =
+        triggers.into_iter().fold(HashMap::new(), |mut map, t| {
+            map.entry(t.block_number()).or_default().push(t);
+            map
+        });
+
+    debug!(logger, "Found {} relevant block(s)", block_hashes.len());
+
+    // Make sure `to` is included, even if empty.
+    block_hashes.insert(to_hash);
+    triggers_by_block.entry(to).or_insert(Vec::new());
+
+    let mut blocks = adapter
+        .load_blocks(logger1, chain_store, block_hashes)
+        .and_then(
+            move |block| match triggers_by_block.remove(&block.number()) {
+                Some(triggers) => Ok(EthereumBlockWithTriggers::new(
+                    triggers,
+                    BlockFinality::Final(block),
+                )),
+                None => Err(format_err!(
+                    "block {:?} not found in `triggers_by_block`",
+                    block
+                )),
+            },
+        )
+        .collect()
+        .compat()
+        .await?;
+
+    blocks.sort_by_key(|block| block.ethereum_block.number());
+
+    // Sanity check that the returned blocks are in the correct range.
+    // Unwrap: `blocks` always includes at least `to`.
+    let first = blocks.first().unwrap().ethereum_block.number();
+    let last = blocks.last().unwrap().ethereum_block.number();
+    if first < from {
+        return Err(format_err!(
```

Comment:
I am not all that familiar with this code - what happens when we return an `Err` here? Does the subgraph fail? It might be better to trim `blocks` to the range we asked for and continue with that, and log an error about what we had to trim because it was outside the requested range.

Reply:
I think it this case we should fail loudly, since it's an inconsistency in the Ethereum node response. Trimming would make sense but then we're likely to miss it if this ever happens.